include/pybind11/cast.h

```diff
@@ -2092,7 +2092,7 @@ class unpacking_collector {
     }
 
     void process(list &args_list, detail::args_proxy ap) {
-        for (const auto &a : ap)
+        for (auto a : ap)
```

Comment:
In case anyone else wonders "why don't we fix the iterator class", here's why:

- "Iterator" is [this thing](https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L836)
- [`operator*`](https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L857-L863) returns a [`py::handle`](https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L841) which is generally better to copy than to take by reference.
- The iterator stores and mutates `value`, so returning a reference here would be wrong as soon as you advance the iterator.

So this is complete fine.

Reply:
Meep! I guess I haven't looked what `const_iterator` looks for non-random-access containers in STL :sweat_smile:

From my understanding, your statement of "why don't we fix the iterator class" means "why don't we change `operator*` return `[const] handle&`.

I think there's still question like "what do other libs do for similar patterns?", especially for iterators implemented for something like `std::list`, etc.

Reply:
(No action here, to be clear, just wanted to clarify on the comment)